default/AI/ResearchAI.py

```diff
@@ -20,21 +32,402 @@ def get_research_index():
+        return 1.5
+    else:
+        return 0.1
+
+def get_xeno_genetics_priority():
+    if foAI.foAIstate.aggression < fo.aggression.cautious:
+        return get_population_boost_priority()
+    if has_only_bad_colonizers():
+        # Empire only have lousy colonisers, xeno-genetics are really important for them
+        print "Empire has only lousy colonizers, increase priority to xeno_genetics"
+        return get_population_boost_priority() * 3
+    else:
+        # TODO: assess number of planets with Adequate/Poor planets owned or considered for colonies
+        return 0.6 * get_population_boost_priority()
+
+def get_artificial_black_hole_priority():
+    if has_star(fo.starType.blackHole) or not has_star(fo.starType.red):
+        print "Already have black hole, or does not have a red star to turn to black hole. Skipping ART_BLACK_HOLE"
+        return 0
+    for tech in AIDependencies.SHIP_TECHS_REQUIRING_BLACK_HOLE:
+        if tech_is_complete(tech):
+            print "Solar hull is researched, needs a black hole to produce it. Research ART_BLACK_HOLE now!"
+            return 999
+    return 1
+
+def get_hull_priority(tech_name):
+    hull = 1
+    offtrack_hull = 0.05
+
+    chosen_hull = choices['hull']
+    organic = hull if chosen_hull % 2 == 0 or choices['extra_organic_hull'] else offtrack_hull
+    robotic = hull if chosen_hull % 2 == 1 or choices['extra_robotic_hull'] else offtrack_hull
+    if ColonisationAI.got_ast:
+        extra = choices['extra_asteroid_hull']
+        asteroid = hull if chosen_hull == 2 or extra else offtrack_hull
+        if asteroid == hull and not extra:
+            organic = offtrack_hull
+            robotic = offtrack_hull
+    else:
+        asteroid = 0
+    if has_star(fo.starType.blue) or has_star(fo.starType.blackHole):
+        extra = choices['extra_energy_hull']
+        energy = hull if chosen_hull == 3 or extra else offtrack_hull
+        if energy == hull and not extra:
+            organic = offtrack_hull
+            robotic = offtrack_hull
+            asteroid = offtrack_hull
+    else:
+        energy = 0
+
+    useful = max(
+        get_ship_tech_usefulness(tech_name, ShipDesignAI.MilitaryShipDesigner()),
+        get_ship_tech_usefulness(tech_name, ShipDesignAI.StandardTroopShipDesigner()),
+        get_ship_tech_usefulness(tech_name, ShipDesignAI.StandardColonisationShipDesigner()))
+
+    if foAI.foAIstate.misc.get('enemies_sighted', {}):
+        aggression = 1
+    else:
+        aggression = 0.1
+
+    if tech_name in AIDependencies.ROBOTIC_HULL_TECHS:
+        return robotic * useful * aggression
+    elif tech_name in AIDependencies.ORGANIC_HULL_TECHS:
+        return organic * useful * aggression
+    elif tech_name in AIDependencies.ASTEROID_HULL_TECHS:
+        return asteroid * useful * aggression
+    elif tech_name in AIDependencies.ENERGY_HULL_TECHS:
+        return energy * useful * aggression
+    else:
+        return useful * aggression
```

Comment:
You mean "should not" right?  I would agree, both for how we normally use aggression (AI aggression level) and for how it is used here (a weighting derived from whether enemies have been seen)  This "* aggression" here should simply come out at this spot.  I would propose renaming 'aggression' to 'enemy_concern' or somesuch, and moving it farther up, and then using it in the calc for 'useful', which I'll go comment on at its spot.